pallets/loans-ref/src/lib.rs

```diff
@@ -679,16 +688,38 @@ pub mod pallet {
 			})
 		}
 
-		fn find_write_off_state(
+		/// From all overdue write off rules, it returns the one has highest percentage
+		/// (or highest penalty, if same percentage) that can be applied.
+		///
+		/// Suppose a policy with the following rules:
+		/// - overdue_days: 5,   percentage 10%
+		/// - overdue_days: 10,  percentage 30%
+		/// - overdue_days: 15,  percentage 20%
+		///
+		/// If the loan is not overdue, it will not return any rule.
+		/// If the loan overdue by 4 days, it will not return any rule.
+		/// If the loan is overdue by 9 days, it will return the first rule.
+		/// If the loan is overdue by 60 days, it will return the second rule
+		/// (because it has a higher percetage).
+		fn find_write_off_rule(
 			pool_id: PoolIdOf<T>,
-			maturity_date: Moment,
-		) -> Result<WriteOffState<T::Rate>, DispatchError> {
-			WriteOffState::find_best(
-				WriteOffPolicy::<T>::get(pool_id).into_iter(),
-				maturity_date,
-				T::Time::now().as_secs(),
-			)
-			.ok_or_else(|| Error::<T>::NoValidWriteOffState.into())
+			loan: &ActiveLoan<T>,
+		) -> Result<Option<WriteOffRule<T::Rate>>, DispatchError> {
+			Ok(WriteOffPolicy::<T>::get(pool_id)
+				.into_iter()
+				.filter_map(|rule| {
+					rule.triggers
+						.iter()
+						.map(|trigger| loan.check_write_off_trigger(&trigger.0))
```

Comment:
Nit: Maybe it would be better to destructure into the 0th tuple element directly to make it more explicit?

```suggestion
						.map(|(trigger, _)| loan.check_write_off_trigger(&trigger))
```

Reply:
It has only one element, the `map` part should be something like `.map(|UniqueWriteOffTrigger(trigger)|)`, which maybe is more confusing (line breaks and adds a new indentation). It if was a tuple, I'd agree with you.